### PR TITLE
Add Engine A/C Correction-Reason Trend Visualizations and Notebook Updates

### LINKED ISSUE #59: Add Engine A & Engine C Visualization Dashboards Aligned with Engine B Reporting
### 🎯 Objective
- **What problem does this Issue solve?**  
  Engine A (reconciliation) and Engine C (Roth taxable) lack standardized visualization outputs, while Engine B has a full visualization module and notebook. This limits consistent KPI reporting and insight across all engines.
- **What is the expected output/deliverable?**  
  New visualization modules and notebooks for Engine A and Engine C that mirror Engine B’s visualization structure and plotting patterns, while adding engine-specific analyses based on each engine’s outputs.

---
### 📌 Scope
**In scope**
- [x] Add visualization modules for Engine A and Engine C that mirror the Engine B structure in `src/age_taxcode_visualization.py`.
- [x] Provide KPI summaries and correction-rate metrics for Engine A and Engine C.
- [x] Add engine-specific charts reflecting each engine’s unique outputs:
  - Engine A: match status distribution, unmatched counts, date lag/tolerance insights, correction reason or distribution category breakouts.
  - Engine C: correction action mix (UPDATE vs INVESTIGATE), Roth taxable correction drivers (basis coverage, qualified age/year), tax code corrections, and taxable amount deltas.
- [x] Add tests for new visualization builders similar to `tests/visualization/test_age_taxcode_visualization.py`.
- [x] Add notebook(s) demonstrating Engine A and C visualizations (similar to `notebooks/06_age_taxcode_visualization.ipynb`).
- [x] Update README to list new visualization modules/notebooks.

**Out of scope**
- Changes to underlying engine business logic (matching, tax code, Roth taxable rules).
- Modifying sample data or production exports.

---
### 🧩 Implementation Plan
**Files to change / add**
- [x] `src/engine_a_visualization.py` *(name tbd: follow existing naming style, e.g., `match_visualization.py`)*
- [x] `src/roth_taxable_visualization.py`
- [x] `notebooks/08_engine_a_visualization.ipynb` *(or append to existing engine-related notebooks)*
- [x] `notebooks/09_roth_taxable_visualization.ipynb`
- [x] `tests/visualization/test_engine_a_visualization.py`
- [x] `tests/visualization/test_roth_taxable_visualization.py`
- [x] `README.md`

**High-level steps**
1. Mirror Engine B visualization structure (metrics builder + KPI summary + plot helpers) for Engine A and Engine C outputs.
2. Add engine-specific aggregation functions:
   - Engine A: match_status groupings (no_action/needs_correction/date_out_of_range/unmatched), date lag distributions, unmatched source counts, correction rates by dist_category_relius.
   - Engine C: action distribution (UPDATE_1099/INVESTIGATE), correction_reason summary, taxable delta distributions, tax-code correction breakdowns, qualified vs non-qualified Roth status rates.
3. Add plotting helpers that follow Engine B’s style (matplotlib, bar/line combos).
4. Add tests for builder functions (required columns validation, empty dataset behavior).
5. Add notebooks demonstrating outputs on sample data.
6. Update README to include the new visualization modules and notebook references.

---
### 🔧 Branch + Commit Plan
**Branch**
- `feature/engine-a-c-visualizations`

**Commits (planned)**
1. `feat: add engine A visualization builders and plots`
2. `feat: add engine C visualization builders and plots`
3. `test/docs: add visualization tests and notebooks`

---
### 🧪 Validation Plan (Local)
**Smoke checks**
- [x] `python -c "from src.match_visualization import build_match_kpi_summary"` → imports without error  
- [x] `python -c "from src.roth_taxable_visualization import build_roth_kpi_summary"` → imports without error  
- [x] Notebook cells in new visualization notebooks run without error

**Data quality checks**
- [x] Match-status KPIs sum to total records (Engine A + C).
- [x] Correction rates are within 0–100%.
- [x] Empty inputs return well-formed empty frames with expected columns.

**Export checks (if applicable)**
- [ ] N/A (visualization-only)

---
### ✅ Acceptance Criteria
- [x] AC1: Engine A visualization module includes KPI summary and at least two engine-specific plots (match status + date lag/unmatched analysis).
- [x] AC2: Engine C visualization module includes KPI summary and at least two engine-specific plots (action mix + taxable/tax-code correction analysis).
- [x] AC3: New tests cover required-column validation and empty input behavior.
- [x] AC4: New notebooks run end-to-end with sample data and render charts.
- [x] AC5: README lists the new visualization modules and notebooks.

---
### 🧯 Risks / Edge Cases
- **Potential risk:** Engine A/C outputs may have missing columns in some runs.
- **Edge case:** Missing txn_date/date_lag_days or missing Roth basis fields could cause plot errors.
- **Mitigation:** Validation helpers for required columns + graceful empty/NA handling (align with Engine B patterns).

---
### 📎 PR Checklist
When opening the PR:
- [x] Include summary + rationale
- [x] Add verification steps (commands/notebook cells)
- [x] Confirm no real data committed
- [x] Link Issue: `Closes #<issue-number>`

### PR DESCRIPTION
**Pull request by manuel-reyes-ml**
# ✅ PR Summary
## 🎯 Objective
**What problem does this PR solve?**  
Add correction-reason trend visualizations for Engine A (reconciliation) and Engine C (Roth taxable) to align with existing KPI reporting and improve consistency across engines.

**Expected output / deliverable**  
New trend builders and plots for correction reasons (month-over-month) in Engine A and C, plus notebook updates and README mention.

---
### In scope
- [x] Add correction-reason trend builders and plots for Engine A.
- [x] Add correction-reason trend builders and plots for Engine C (split multi-line reasons).
- [x] Update Engine A/C visualization notebooks to include trend outputs.
- [x] Update README to list the new trend visualization.

### Out of scope
- Changes to engine business logic or matching/taxable rules.
---
### Files changed / added
- [x] `src/match_planid_visualization.py`
- [x] `src/roth_taxable_visualization.py`
- [x] `notebooks/08_match_planid_visualization.ipynb`
- [x] `notebooks/09_roth_taxable_visualization.ipynb`
- [x] `README.md`

### High-level approach
1. Add correction-reason trend builders with required-column validation and empty-frame handling.
2. Add trend plot helpers that visualize monthly counts by correction reason.
3. Update notebooks and README to surface the new trend outputs.

---
### Business rules implemented / updated
- **Rule(s):** Build month-over-month counts by correction_reason for records needing correction (Engine A) or correction/review (Engine C).
- **Threshold(s):** N/A
- **Exclusions / locks:** Ignore rows without valid txn_date; Engine C ignores empty correction_reason after splitting.

### Canonical schema impact
- **New columns added:** None
- **Columns modified:** None
- **No schema change:** [x]

### Data quality considerations
- **Join keys:** N/A
- **Null-handling:** Missing/malformed txn_date raises ValueError; missing correction_reason yields empty metrics.
- **Type enforcement:** txn_date coerced to datetime; correction_reason cast to string and split for Engine C.
- **Idempotence:** Yes; metrics are derived from input data.

---
### Smoke checks
- [x] `python -c "from src.match_planid_visualization import build_correction_reason_trends"` runs without error
- [x] `python -c "from src.roth_taxable_visualization import build_roth_correction_reason_trends"` runs without error
- [x] Notebook cell(s) run without error

### Data quality checks
- [x] Trend metrics return empty frames with expected columns on empty inputs
- [x] txn_date validation rejects malformed or missing dates
- [x] Correction reason splits are counted correctly (Engine C)

### Validation (executed)
```bash
python -c "from src.match_planid_visualization import build_correction_reason_trends"
python -c "from src.roth_taxable_visualization import build_roth_correction_reason_trends"
```
```bash
python -m pytest tests/visualization/test_match_planid_visualization.py tests/visuali
zation/test_roth_taxable_visualization.py
```

**Results:** 
    - Completed succesfully without syntax error.
    - Pytest: 29 passed in 1.00s

### Rule verification (recommended)
- [x] Added/updated notebook examples for correction-reason trend outputs

### Export checks (if applicable)
N/A

---
## ✅ Acceptance Criteria
- [x] AC1: Engine A includes correction-reason trend builder + plot.
- [x] AC2: Engine C includes correction-reason trend builder + plot (multi-line reason parsing).
- [x] AC3: Notebooks updated to demonstrate new trend outputs.
- [x] AC4: README mentions new trend visualization.

---
## 🧯 Risks / Edge Cases
- Potential risk: Missing or malformed txn_date raises validation errors.
- Edge cases covered: Empty inputs return well-formed empty metrics.
- Mitigation: Explicit required-column checks and input validation.

---
### What to focus on
- Correct handling of txn_date validation and empty frames
- Reason splitting and counting for Engine C
- Notebook outputs and README mention

### Screenshots / sample outputs (optional)
<details> <summary>pytest run output (local)</summary> 
Screenshot 2026-01-01 at 11 17 54 PM

</details>

---
## 🔗 Linking
- Closes #59

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -273,11 +274,13 @@ New First Year contrib | Reason | Action
 │   ├── clean_relius_roth_basis.py      # Relius Roth basis cleaning
 │   ├── clean_matrix.py                 # Matrix cleaning logic
 │   ├── match_transactions.py           # Engine A (inherited matching)
+│   ├── match_visualization.py          # Engine A charts
```
[the remark this answers is not in the document]
Comment (manuel-reyes-ml) [reply README.md]: you are totally right. However, this is intentional for now. We made last minute name changes to files to better describe module vs engines purpose. This is our next issue on list to solve!
